src/optimizer/rule/ordered_aggregate_optimizer.cpp

```diff
@@ -24,6 +30,90 @@ unique_ptr<Expression> OrderedAggregateOptimizer::Apply(LogicalOperator &op, vec
 		changes_made = true;
 		return nullptr;
 	}
+
+	auto &context = rewriter.context;
+	if (context.config.enable_optimizer) {
+		// for each ORDER BY - check if it is actually necessary
+		// expressions that are in the groups do not need to be ORDERED BY
+		// `ORDER BY` on a group has no effect, because for each aggregate, the group is unique
+		// similarly, we only need to ORDER BY each aggregate once
+		expression_set_t seen_expressions;
+		auto &groups = op.Cast<LogicalAggregate>().groups;
+		for (auto &target : groups) {
+			seen_expressions.insert(*target);
+		}
+		vector<BoundOrderByNode> new_order_nodes;
+		for (auto &order_node : aggr.order_bys->orders) {
+			if (seen_expressions.find(*order_node.expression) != seen_expressions.end()) {
+				// we do not need to order by this node
+				continue;
+			}
+			seen_expressions.insert(*order_node.expression);
+			new_order_nodes.push_back(std::move(order_node));
+		}
+		if (new_order_nodes.empty()) {
+			aggr.order_bys.reset();
+			changes_made = true;
+			return nullptr;
+		}
+		aggr.order_bys->orders = std::move(new_order_nodes);
+	}
+
+	//	Rewrite first/last/arbitrary/any_value to use arg_xxx[_null] and create_sort_key
+	const auto &aggr_name = aggr.function.name;
+	string arg_xxx_name;
+	if (aggr_name == "last") {
+		arg_xxx_name = "arg_max_null";
+	} else if (aggr_name == "first" || aggr_name == "arbitrary") {
+		arg_xxx_name = "arg_min_null";
+	} else if (aggr_name == "any_value") {
+		arg_xxx_name = "arg_min";
+	}
+
+	if (!arg_xxx_name.empty()) {
+		FunctionBinder binder(context);
+		vector<unique_ptr<Expression>> sort_children;
+		for (auto &order : aggr.order_bys->orders) {
+			sort_children.emplace_back(std::move(order.expression));
+
+			string modifier;
+			modifier += (order.type == OrderType::ASCENDING) ? "ASC" : "DESC";
+			modifier += " NULLS";
+			modifier += (order.null_order == OrderByNullType::NULLS_FIRST) ? " FIRST" : " LAST";
+			sort_children.emplace_back(make_uniq<BoundConstantExpression>(Value(modifier)));
+		}
+		aggr.order_bys.reset();
+
+		string error;
+		auto sort_key = binder.BindScalarFunction(DEFAULT_SCHEMA, "create_sort_key", std::move(sort_children), error);
+		if (!sort_key) {
+			throw BinderException(error);
+		}
+
+		auto &children = aggr.children;
+		children.emplace_back(std::move(sort_key));
+
+		//  Look up the arg_xxx_name function in the catalog
+		QueryErrorContext error_context(nullptr, 0);
```

Comment:
This can just be `QueryErrorContext error_context` - no need to provide arguments here